controller/cmd/service-mirror/probe_manager.go

```diff
@@ -0,0 +1,194 @@
+	metricVecs := newProbeMetricVecs()
+	return &ProbeManager{
+		k8sAPI:       k8sAPI,
+		probeWorkers: make(map[string]*ProbeWorker),
+		events:       events,
+		metricVecs:   &metricVecs,
+		done:         make(chan struct{}, 1),
+	}
+}
+
+// ProbeManager takes care of managing the lifecycle of probe workers
+type ProbeManager struct {
+	k8sAPI       *k8s.API
+	probeWorkers map[string]*ProbeWorker
+
+	events     chan interface{}
+	metricVecs *probeMetricVecs
+	done       chan struct{}
+}
+
+// GatewayProbeSpec is the specification of a probe that will be executed by the worker
+type GatewayProbeSpec struct {
+	clusterName   string
+	gatewayName   string
+	gatewayNs     string
+	gatewayIps    []string
+	port          int32
+	path          string
+	periodSeconds int32
+}
+
+// MirroredServiceUnpaired is emitted when a service is no longer mirrored
+type MirroredServiceUnpaired struct {
+	serviceName      string
+	serviceNamespace string
+	gatewayName      string
+	gatewayNs        string
+	clusterName      string
+}
+
+// MirroredServicePaired is emitted when a new mirrored service is created
+type MirroredServicePaired struct {
+	serviceName      string
+	serviceNamespace string
+	*GatewayProbeSpec
+}
+
+// ClusterRegistered is emitted when a new cluster becomes registred for mirroring
+type ClusterRegistered struct {
+	clusterName   string
+	port          int32
+	path          string
+	periodSeconds int32
+}
+
+// GatewayUpdated is emitted when something about the gateway is updated (i.e. its external IP(s))
+type GatewayUpdated struct {
+	*GatewayProbeSpec
+}
+
+func probeKey(gatewayNamespace string, gatewayName string, clusterName string) string {
+	return fmt.Sprintf("%s-%s-%s", gatewayNamespace, gatewayName, clusterName)
+}
+
+func (m *ProbeManager) handleEvent(ev interface{}) {
+	switch ev := ev.(type) {
+	case *MirroredServicePaired:
+		m.handleMirroredServicePaired(ev)
+	case *MirroredServiceUnpaired:
+		m.handleMirroredServiceUnpaired(ev)
+	case *GatewayUpdated:
+		m.handleGatewayUpdated(ev)
+	case *ClusterRegistered:
+		m.handleClusterRegistered(ev)
+	}
+}
+
+func (m *ProbeManager) handleMirroredServicePaired(event *MirroredServicePaired) {
+	probeKey := probeKey(event.gatewayNs, event.gatewayName, event.clusterName)
+	worker, ok := m.probeWorkers[probeKey]
+	if ok {
+		log.Debugf("Probe worker %s already exists", probeKey)
+		worker.IncNumServices()
+	} else {
+		log.Debugf("Creating probe worker %s", probeKey)
+		probeMetrics := m.metricVecs.newMetrics(event.gatewayNs, event.gatewayName, event.clusterName)
+		worker = NewProbeWorker(event.GatewayProbeSpec, &probeMetrics, probeKey)
+		worker.IncNumServices()
+		m.probeWorkers[probeKey] = worker
+		worker.Start()
+	}
+}
+
+func (m *ProbeManager) handleMirroredServiceUnpaired(event *MirroredServiceUnpaired) {
+	probeKey := probeKey(event.gatewayNs, event.gatewayName, event.clusterName)
+	worker, ok := m.probeWorkers[probeKey]
+	if ok {
```

Comment:
Would it be helpful to log the case that a service unpairing event occurs, but there is not a worker for that gateway/cluster?